qiskit/dagcircuit/dagnode.py

```diff
@@ -163,26 +184,93 @@ def semantic_eq(node1, node2, bit_indices1=None, bit_indices2=None):
             bit_indices1 = {arg: arg for arg in node1.qargs + node1.cargs}
             bit_indices2 = {arg: arg for arg in node2.qargs + node2.cargs}
 
-        node1_qargs = [bit_indices1[qarg] for qarg in node1.qargs]
-        node1_cargs = [bit_indices1[carg] for carg in node1.cargs]
-
-        node2_qargs = [bit_indices2[qarg] for qarg in node2.qargs]
-        node2_cargs = [bit_indices2[carg] for carg in node2.cargs]
-
-        # For barriers, qarg order is not significant so compare as sets
-        if "barrier" == node1.name == node2.name:
-            return set(node1_qargs) == set(node2_qargs)
-
-        if node1.type == node2.type:
-            if node1._op == node2._op:
-                if node1.name == node2.name:
-                    if node1_qargs == node2_qargs:
-                        if node1_cargs == node2_cargs:
-                            if node1.type == "op":
-                                if node1._op.condition != node2._op.condition:
-                                    return False
-                            if bit_indices1.get(node1._wire, None) == bit_indices2.get(
-                                node2._wire, None
-                            ):
-                                return True
+        if isinstance(node1, OpNode) and isinstance(node2, OpNode):
+            node1_qargs = [bit_indices1[qarg] for qarg in node1.qargs]
+            node1_cargs = [bit_indices1[carg] for carg in node1.cargs]
+
+            node2_qargs = [bit_indices2[qarg] for qarg in node2.qargs]
+            node2_cargs = [bit_indices2[carg] for carg in node2.cargs]
+
+            # For barriers, qarg order is not significant so compare as sets
+            if "barrier" == node1.op.name == node2.op.name:
+                return set(node1_qargs) == set(node2_qargs)
+
+            if node1_qargs == node2_qargs:
+                if node1_cargs == node2_cargs:
+                    if node1.op.condition == node2.op.condition:
+                        if node1.op == node2.op:
+                            return True
+        elif (isinstance(node1, InNode) and isinstance(node2, InNode)) or (
+            isinstance(node1, OutNode) and isinstance(node2, OutNode)
+        ):
+            if bit_indices1.get(node1.wire, None) == bit_indices2.get(node2.wire, None):
+                return True
+
         return False
+
+
+class OpNode(DAGNode):
+    """Object to represent an Instruction at a node in the DAGCircuit."""
+
+    __slots__ = ["op"]
+
+    def __init__(self, op, qargs=None, cargs=None):
+        """Create an Instruction node"""
+        self.op = op
+        super().__init__(qargs=qargs, cargs=cargs)
```

Comment:
Should `qargs` and `cargs` be migrated to `OpNode` if they are only set there (as was done for `wire` for `{In,Out}Node`s)?

Reply:
(As mentioned above) The _qargs are used to create the sort_key which is used in dagcircuit.topological_nodes for all types of nodes. That's why they're handled here by the parent.

Reply:
Ok, I see now. In that case, this is probably a bug since `self._qargs` will always be `[]` for `in` and `out` nodes. Instead, perhaps the `sort_key` for `InNode` and `OutNode` should be `str(self.wire)`.

Reply:
Ah. Good point. Will try replacing with 'self.wire'.

Reply:
I did a quick switch to use 'sort_key = str(self.wire)' for the in and out nodes, and it failed 1 dagcircuit test and 10 transpiler ones. The dagcircuit was the test_topological_nodes test, so it could just be modified. But the transpiler ones were across 5 different passes, and the failures were not obvious. Maybe we deal with changing the sort_key as another issue.

Reply:
I'm assuming we'll leave this as is for now?

Reply:
I can look into the failing cases and see if there's a straightforward resolution. I'd rather not add in new kwargs knowing that they'll have to be deprecated and removed, if we can avoid it.

Reply:
qargs and cargs moved to DAGOpNode and sort_key is sorted. https://github.com/Qiskit/qiskit-terra/pull/6567/commits/92c369385b428bf10fb68fdd8d2fd44f6baa6d90